modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -487,28 +495,32 @@ def downsample_trigger_training_set(self) -> None:
         assert self._downsampler is not None
         assert self._downsampling_mode == DownsamplingMode.SAMPLE_THEN_BATCH
 
+        # set the model to eval to avoid errors like Expected more than 1 value per channel when training, got ...
+        self._model.model.eval()
         # keys must be taken from the selector.
         # This operation is needed only when we sample several times (otherwise the source is already the selector)
         selector_key_source = SelectorKeySource(
             pipeline_id=self.pipeline_id, trigger_id=self.trigger_id, selector_address=self.selector_address
         )
         self._train_dataloader.dataset.change_key_source(selector_key_source)
-
-        number_of_samples = 0
-
         self._downsampler.init_downsampler()
-
-        batch_number = 0
-        for batch_number, batch in enumerate(self._train_dataloader):
-            self.update_queue(AvailableQueues.DOWNSAMPLING, batch_number, number_of_samples, training_active=False)
-
-            sample_ids, target, data = self.preprocess_batch(batch)
-            number_of_samples += len(sample_ids)
-
-            with torch.autocast(self._device_type, enabled=self._amp):
-                # compute the scores and accumulate them
-                model_output = self._model.model(data)
-                self._downsampler.inform_samples(sample_ids, model_output, target)
+        if self._downsampler.requires_data_label_by_label:
+            available_labels = self._get_available_labels_from_selector()
+            per_class_dataloader = prepare_per_class_dataloader_from_online_dataset(
+                self._train_dataloader.dataset, self._batch_size, self._num_dataloaders
+            )
+            number_of_samples = 0
+            batch_number = 0
+            for label in available_labels:
+                per_class_dataloader.dataset._filtered_label = label
```

Comment:
ah now I see. usually a leading underscore indicates a private attribute. we should pass an initial filtered label in the constructor and make this attribute public

Reply:
Ok